[![PyPI](https://img.shields.io/badge/Install%20with-PyPI-blue)](https://pypi.org/project/sequenoscope/#description)
[![Bioconda](https://img.shields.io/badge/Install%20with-bioconda-green)](https://anaconda.org/bioconda/sequenoscope)
[![Conda](https://img.shields.io/conda/dn/bioconda/profile_dists?color=green)](https://anaconda.org/bioconda/sequenoscope)
[![License: Apache-2.0](https://img.shields.io/github/license/phac-nml/profile_dists)](https://www.apache.org/licenses/LICENSE-2.0)


## Sequenoscope

<img height="150" width="400" alt="logo11" src="https://user-images.githubusercontent.com/93303799/225326096-6c9de0f1-9ac0-46a4-914f-a3db51e7e97a.png">

A tool for analyzing sequencing output. 

## Contents

- [Introduction](#introduction)
- [Installation](#installation)
- [Usage](#usage)
- [Quick Start](#quick-start)
- [FAQ](#faq)
- [Citation](#citation)
- [Legal](#legal)
- [Contact](#contact)

## Introduction

Analyzing and interpreting sequencing data is a fundamental task in bioinformatics, especially when studying adaptive sampling MinION sequencing data. To address this challenge, we have developed a comprehensive bioinformatics pipeline consisting of three modules: "analyze," "plot," and "filter_ONT." Our pipeline aims to provide researchers with a streamlined and efficient workflow for processing and analyzing sequencing data, enabling them to gain valuable insights into their datasets.

The "analyze" module serves as the core component of our pipeline. It takes an input fastq file, a reference fastq file, and an optional sequencing summary file from ONT sequencers. Leveraging tools such as fastp, minimap2, pysam, and kat, this module performs a series of essential tasks. It filters the input fastq file, maps it to the reference fastq file, and generates sequence manifest files and summary sequence manifest files, which contain various sequencing statistics. Additionally, it produces informative kmer plots using the kat program.

The "plot" module complements the analysis performed by the "analyze" module. It takes as input both test and control sequence manifest or a sequence manifest summary files generated by the "analyze" module. With these files, the "plot" module generates visualizations that aid in the interpretation and visualization of the sequencing data.

To further enhance the versatility of our pipeline, we have developed the "filter_ONT" module. This module enables users to filter reads from a fastq file based on a sequencing summary file. Researchers can customize their filtering criteria by specifying parameters such as channel, duration, start time, q score, and length.

By providing a comprehensive suite of modules, our bioinformatics pipeline offers a powerful tool for researchers working with adaptive sampling MinION sequencing data. Whether you are exploring metagenomics sample composition, investigating adaptive sampling for your project, or conducting a comparative analysis of different methods in your lab, our pipeline can streamline your analyses and provide valuable insights into your genomic datasets.

## Installation

Install the latest released version from conda:

        conda create -c bioconda -c conda-forge -n sequenoscope

**Coming soon**

Install using pip:

        pip install sequenoscope

**Coming soon**

Install the latest master branch version directly from Github:

        pip install git+https://github.com/phac-nml/sequenoscope.git



## Usage
If you run ``sequenoscope``, you should see the following usage statement:

        Usage: sequenoscope <command> <required arguments>
        
        To get full help for a command use one of:
        sequenoscope <command> -h
        sequenoscope <command> --help
        
        
        Available commands:
        
        analyze     map reads to a target and produce a report with sequencing statistics
        plot        generate plots based on fastq or kmer hash files
        filter_ONT  filter reads from a fastq file based on a sequencing summary file

Quick start
=====
To quickly get started with the `analyze` module, follow the instructions below:

1. Ensure that you have the necessary input files and reference database prepared:
   - **Input Fastq files:** Provide the path to the Fastq files you want to process using the `--input_fastq` option.
   - **Reference database:** Specify the path to the reference database using the `--input_reference` option.

2. Choose an output directory for the results:
   - Specify the output directory using the `--output` option. This is a required parameter.

3. Run the command with the basic required options:

        sequenoscope analyze --input_fastq <file.fq> --input_reference <ref.fasta> -o <output_directory>

This command will initiate the analysis process using default settings. The Fastq files will be processed, and the results will be saved in the specified output directory.

Please note that this is a simplified quick start guide, and additional options are available for advanced usage. For more detailed information on available options, you can run `sequenoscope analyze -h` or `sequenoscope analyze --help`.

Remember to replace `<file.fq>` with the actual path to your Fastq file, `<ref.fasta>` with the path to your reference database, and `<output_directory>` with the desired location for the output files.

If you encounter any issues or need further assistance, refer to the full documentation or consult the available resources for troubleshooting.

-------

To quickly get started with the `filter_ONT` module, follow the instructions below:

1. Ensure that you have the necessary input files prepared:
   - **Input Fastq files:** Provide the path to the adaptive sequencing Fastq files you want to process using the `--input_fastq` option.
   - **ONT sequencing summary file:** Specify the path to the ONT sequencing summary file using the `--input_summary` option.

2. Choose an output file and directory for the filtered reads:
   - Specify the output file path and directory using the `--output` option. This is a required parameter.

3. Set the desired filtering criteria:
   - You can apply various filters to the reads based on the following options:
     - Classification: Use the `-cls` or `--classification` option to designate the adaptive-sampling sequencing decision classification. Valid options are `'unblocked'`, `'stop_receiving'`, or `'no_decision'`.
     - Channel/Pore number: Set the minimum and maximum channel/pore number for filtering using the `-min_ch` and `-max_ch` options.
     - Duration: Define the minimum and maximum duration of the sequencing run in seconds using the `-min_dur` and `-max_dur` options.
     - Start time: Specify the minimum and maximum start time of the sequencing run in seconds using the `-min_start` and `-max_start` options.
     - Q score: Determine the minimum and maximum q score for filtering using the `-min_q` and `-max_q` options.
     - Read length: Set the minimum and maximum read length for filtering using the `-min_len` and `-max_len` options.

4. Run the command with the basic required options:

        sequenoscope filter_ONT --input_fastq <file.fq> --input_summary <seq_summary.txt> -o <output.fastq>

This command will initiate the filtering process based on the specified criteria and save the filtered reads to the output file.

Please note that this is a simplified quick start guide, and additional options are available for advanced usage. For more detailed information on available options, you can run `sequenoscope filter_ONT -h` or `sequenoscope filter_ONT --help`.

Remember to replace `<file.fq>` with the actual path to your adaptive sequencing Fastq file, `<seq_summary.txt>` with the path to your ONT sequencing summary file, and `<output.fastq>` with the desired path and filename for the filtered reads.

If you encounter any issues or need further assistance, refer to the full documentation or consult the available resources for troubleshooting.

-------------
To quickly get started with the `plot` module, follow the instructions below:

**Coming soon**

## Benchmarks

**Coming soon**

## FAQ

**Coming soon**

## Citation

**Coming soon**

## Legal

Copyright Government of Canada 2023

Written by: National Microbiology Laboratory, Public Health Agency of Canada

Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this work except in compliance with the License. You may obtain a copy of the
License at:

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed
under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.


## Contact

**Abdallah Meknas**: abdallah.meknas@phac-aspc.gc.ca

